packages/library/src/commonMain/kotlin/io/realm/RealmResults.kt

```diff
@@ -30,29 +30,70 @@ import kotlin.reflect.KClass
 //  - Lazy API makes it harded to debug
 //  - Postponing execution to actually accessing the elements also prevents query parser errors to
 //    be raised. Maybe we can get an option to prevalidate queries in the C-API?
-class RealmResults<T : RealmObject> constructor(
-    private val realmConfiguration: RealmConfiguration,
-    private val realm: NativePointer,
-    private val queryPointer: () -> NativePointer,
-    private val clazz: KClass<T>,
-    private val mediator: Mediator
-) : AbstractList<T>(), Queryable<T> {
+class RealmResults<T : RealmObject> : AbstractList<T>, Queryable<T> {
+
+    private val mode: Mode
+    private val owner: BaseRealm
+    private val realm: NativePointer // Store explicit reference to pointer because the owner Realm might replace it.
+    private val clazz: KClass<T>
+    private val schema: Mediator
+    private val query: NativePointer
+    private val result: NativePointer
+
+    private enum class Mode {
+        // FIXME: Needed to make working with @LinkingObjects easier.
+        EMPTY, // RealmResults that is always empty.
+        QUERY, // RealmResults created from a query
+        RESULTS // RealmResults wrapping a Realm Core Results.
+    }
+
+    // Create Results from query
+    private constructor(realm: BaseRealm, query: () -> NativePointer, clazz: KClass<T>, schema: Mediator) {
+        this.mode = Mode.QUERY
+        this.owner = realm
+        this.realm = realm.dbPointer
+        this.query = query()
```

Comment:
If we are always evaluating this we should maybe just supply the native pointer from outside directly. The only need for this is, as far as I remember, to get the `result` pointer so maybe we could skip the `query` and only track the `result`. Might also impact when the majority of the load is performed, but probably have to have some impact from core team to reassess. 

Might be that moving it around and thawing it need the query!? 🤔

Reply:
I assume the old behavior was to lazy evaluate the query as late as possible, but with this architecture, it probably doesn't make sense and we can just evaluate it directly.

But yes, there is probably room for optimization here.

Reply:
Hmm, thinking about this. We probably need to be a bit careful how this is used. Evaluating the query itself is fairly cheap I believe since it just validates the query itself, but e.g. running a `find_all` is definitely not and is something we would want to move to the background thread if at all possible.

Reply:
Turns out that our `realm_query_find_all` is lazy. So it doesn't actually run the query until you interact with it. This seems to be the semantics we want, but there isn't a need to store the query pointer itself inside our `RealmResults`...we just need to wrap the underlying Results class which takes care of the rest.